File: cpp/Arrays/GroupAnagram.cpp
//
// Created by Keanu Lumpkin on 5/2/23.
//
#include <string>
#include <vector>
#include <unordered_map>
#include <map>
#include <iostream>

using namespace std;

class GroupAnagram {
public:
    static
    vector<vector<string>> solution(vector<string> &strs) {
        vector<vector<string>> ans;
        unordered_map<string, vector<string>> map;
        for (auto i: strs) {
            string sorted_string = i;
            sort(sorted_string.begin(), sorted_string.end());
            map[sorted_string].push_back(i);
        }
        for (auto i: map) {
            ans.push_back(i.second);
        }
        return ans;
    }

private:
    void printList() {

    }
};

File: csharp/Arrays/GroupAnagram.cs
using System.Collections.ObjectModel;

namespace csharp.Arrays;

public class GroupAnagram
{
    public static IList<IList<string>> Solution(string[] strs)
    {
        Dictionary<string, List<string>> dictionary = new Dictionary<string, List<string>>();
        IList<IList<string>> ans = new List<IList<string>>();

        foreach (var s in strs)
        {
            char[] charKey = s.ToCharArray();
            Array.Sort(charKey);
            string i = new string(charKey);
            if (!dictionary.ContainsKey(new string(charKey)))
            {
                dictionary.Add(i, new List<string>());
            }

            dictionary[i].Add(s);
        }

        foreach (var pair in dictionary)
        {
            ans.Add(pair.Value);
        }

        return ans;
    }
}